README
DiscSpan 0.2.2

author:  James S. Martin  ceedvd ( a t )  g  m a i l .com
maintainer: Richard M. Shaw  hobbes1069 ( a t ) g m a i l .com


---------

This tool is geared more towards smaller files such as music and photos 
as it cannot span a single file across multiple discs.  

If you have lots of files in the gigabyte  range, this utility won't be as efficient.

The tool is written in Python and uses growisofs as the burning interface.


Requirements
------------

It is known to work in Fedora 12, 13, RHEL 5.5, but should work on any recent 
Linux system with growisofs, and python 2.5-2.6.  Requires the dbus-python module.


running and installing DiscSpan
----------------

copy the discscan.py somewhere in your path.  Put discpan.ini in /etc, 
/usr/local/etc, the same dir as discspan.py, or specify with --config=

Please edit discspan.ini to reflect the speed of your media/device,
as well as the capacity of your media.  I have seen cases where 
media sizes are slightly varied


run discspan.py.  

Usage: discspan.py [options]

Options:
  -h, --help            show this help message and exit
  --config=CONFIG_FILE  Location of config file.
  --start-disc=START_DISC
                        Specify disc to start with (in case of failed previous
                        burn)
  --skip-big            Skip files that are too big.
  --test                Performs a test run of the burn.
  --dir=BACKUP_DIR      Directory to backup.
  --volume-name=VOLUME_NAME
                        Name for the volume.
  --size-factor=SIZE_FACTOR
                        Specify size factor for disc capacity, i.e. 1 = 100%
  --iso-dir=ISO_DIR     Redirect iso generation to a directory. Filename will
                        be generated from volume name.
  -v, --verbose         Extended verbosity



If you don't supply a backup dir, you will be prompted.



limitations
-----------

The discspan.ini measurements are defined in gibibyte and mebibyte.

1G is defined as 2^30 bytes.

Lots of media manufactures define their stuff in gigabytes, which is
93% less than a gibibyte.  

That's why the nubmers you see in discspan.ini may look different
than what's labeled on your media.  Feel free to tweak to your 
hearts content.



Read here:  

http://en.wikipedia.org/wiki/DVD-R
http://en.wikipedia.org/wiki/Gigabyte




discspan will not follow symlinks (yet) !

DBUS doesn't detect the capacity of the disc properly so I can't modify
these on the fly.  If someone can give me the capacity in bytes of a Dual Layer
DVD or Blue Ray/HD-DVD, I can add that.

I've guestimated values but if they are wrong, please tweek them and send
me an appropriate update.

Also, this has only been tested on dvd-r and dvd-rw media.




Thank Yous
-----------
Thanks go out to Doc, Richard, Dez, Erin, Jason, Adam for their debugging, code contributions, and ideas!


Thank you to Davyd Madeley and Dieter Verfaillie for their DBUS
code examples!  Without that, I would not have figured out
how to detect the DVD drive :).

Also thanks to the #python channel on freenode.  Wonderful folks.


CHANGELOG
DiscSpan Changelog:

TODO:

Support for following symlinks
WARNING - If your stuff has symlinks, they will NOT be copied.

Strip out the depedencie on HAL for getting disc / drive info,
and put in something that's not on the list to be deprecated.

Work with python burn developer to get disc spanning feature into
their product. http://pypi.python.org/pypi/burn/0.4.6

Win the lottery; build a hotel on the Moon; and retire.

*** 0.2.2 - 2010.08.09 ***

Fixed: Option "--test" option was broken in 0.2.1.

Fixed: Don't eject tray when --test or --iso-dir options are used.

Fixed: Cleaned up drive/media detection code a little and added more meaninful output. NOTE: On my system HAL does not get updated properly and causes issues detecting media. 

Added: Option "--skip-big" was added to optionally skip files that were too big for ISO 9660 or recording media (CDs) instead of exiting.

Added: Logic to try to predict ISO file system size to reduce need for --size-factor option.


*** 0.2.1 - 2010.07.29 ***

* "--size-factor" option was added for modifying the capacity of the disk on the fly. 
* "--verbose" option will spit out the file names which is now normally suppresed. 
* "--iso-dir" option was added to write to ISO files instead of directly to the burner
* Some more diagnostic output was added, such as how many files will be on each disc and the approximate size of the finished disc. 


*** 0.2.0 - 2010.06.08 ***

Mostly re-written with less suck.  Tried an object-oriented approach.

Should work on modern systems, tested on Fedora 12.

Requires a discspan.ini file, speed _must_ be set in this file. 

Have added a --help option, and using real option parsing now.

We are using isolevel 4 for burns, as to not run in to 
directory depth problems that exist with isolevel 3.

Default config files can be /etc/discspan.ini or /usr/local/etc/discspan.ini.
Otherwise it must be specified with the below option.

New options:

Usage: discspan.py [options]

Options:
  -h, --help            show this help message and exit
  --start-disc=START_DISC
                        Specify disc to start with (in case of failed previous
                        burn)
  --config=CONFIG_FILE  Location of config file.
  --test                Performs a test run of the burn.
  --dir=BACKUP_DIR      Directory to backup.
  --volume-name=VOLUME_NAME
                        Name for the volume.

If you leave out --dir, you will be prompted for it.  The volume name
will default to "spandisc" if it is not specified.


*** 0.1.2 - 2007.12.3 ***

Now capable of detecting dvd_rw, and dvd+ variants.
Correctly traps CTRL+C.
updated README.


*** 0.1.1 - 2007.12.12 ***

Better detection of disc media


*** 0.1  - 2007.12.12 ***

Initial Release



discspan.py
#! /usr/bin/env python

__author__ = "James S. Martin"
__maintainer__ = "Richard M. Shaw"
__credits__ = "Doc, Dez, Erin, Jason, Adam"
__license__ = "GPL"
__version__ = "0.2.2"

import os
import sys
import os.path
import math
import subprocess
import tempfile
import dbus
import readline
import ConfigParser
import string
from decimal import *
from optparse import OptionParser
from time import sleep

debug = False

class Config:

    def __init__(self, cfgfile):

        self.config = ConfigParser.SafeConfigParser()
        self.config.read(cfgfile)
        self.media = self.config.options('media')
        self.speed = self.config.get('drive', 'speed')

    def convert_to_bytes(self,value):

        if value[-1:] == 'M':
            s_value = value.replace('M','')
            b_value = Decimal(s_value) * 2**20

        elif value[-1:] == 'G':
            s_value = value.replace('G','')
            b_value = Decimal(s_value) * 2**30

        else:
            print "Size does matter. Make sure you've specified a M or G after your media config file."
            sys.exit(1)

        return str(b_value)

    def get_capacity(self, medium):
        "Returns capacity of disc type in both human readable and in bytes."

        h_capacity, b_capacity = string.split(self.config.get('media', medium))

        return h_capacity, b_capacity


class Drive:

    def __init__(self, device_name, capacity, drive_name):

        self.device = device_name
        self.capacity = capacity
        self.model = drive_name

class System:

    def __init__(self, config):

        self.config = config

    def find_drives(self):

        drive = None
        bus = dbus.SystemBus()
        hal_obj = bus.get_object('org.freedesktop.Hal', '/org/freedesktop/Hal/Manager')
        hal = dbus.Interface(hal_obj, 'org.freedesktop.Hal.Manager')
        drives = []
        udis = []

        for udi in hal.FindDeviceByCapability('volume'):
            if debug: print udi

            dev_obj = bus.get_object('org.freedesktop.Hal', udi)
            dev = dbus.Interface(dev_obj, 'org.freedesktop.Hal.Device')

            try: volume_disc_type = dev.GetProperty('volume.disc.type')
            except: continue

            if volume_disc_type in self.config.media:
                device_name = dev.GetProperty('block.device')
                print "Media found in %s" % device_name

            else:
                print "No recordable media found in %s." % device_name
                continue

            if not dev.GetProperty('volume.disc.is_blank'):
                print "HAL says media in %s is not blank." % dev.GetProperty('block.device')
                continue

            else:
                parent_obj = bus.get_object('org.freedesktop.Hal', dev.GetProperty("info.parent"))
                parent = dbus.Interface(parent_obj, 'org.freedesktop.Hal.Device')
                h_capacity , b_capacity = self.config.get_capacity(dev.GetProperty('volume.disc.type'))
                drive_name = parent.GetProperty('info.product')
                drive = Drive(device_name, b_capacity, drive_name)
                print "The disc capacity of the disc in %s (%s) is %s." % (drive.device, drive.model, h_capacity)
                drives.append(drive)

        if len(drives) == 1:
            drive = drives[0]
            statement = '\nUsing %s (%s) as your dvd burner' % (drive.device, drive.model)

        elif len(drives) > 1:
            print "Don't make this difficult.  Please only put one piece of writable" \
            "medium in your system at a time."
            sys.exit(1)

        return drive

    def wait_for_media(self):

        n = 0
        drive = self.find_drives()
        if drive == None:
            print "Insert disc and wait for drive to become ready..."
            
        while drive == None:
            drive = self.find_drives()
            sleep(3)
            n = n + 1
            if n >= 30:
                print "Unable to detect media, exiting."
                sys.exit(1)

        return drive

class Interface:

    def __init__(self, backup_dir, speed):

        if self.validate_speed(speed):
            self.speed = speed
        else:
            print 'Your drive speed (%s) is invalid.' % self.speed
            sys.exit(1)

        if backup_dir == None:
            self.backup_dir = self.ask_questions()
        elif self.validate_backup_dir(backup_dir):
            self.backup_dir = backup_dir
        else:
            print 'Directory option is not valid.'
            self.backup_dir = self.ask_questions()

    def ask_questions(self):

        valid = False
        while not valid:
            backup_dir = raw_input('Which directory would you like to backup? (Or Q to quit)\n')
            if backup_dir in ('Q','q'):
                sys.exit(1)
            valid = self.validate_backup_dir(backup_dir)

        self.backup_dir = backup_dir

        return self.backup_dir

    def validate_speed(self, speed):

        try:
            int(speed)
            return True
        except:
            print speed, 'is not a valid number.'
            return False

    def validate_backup_dir(self, backup_dir):

        if os.path.isdir(backup_dir):
            return True
        else:
            print "You must enter a valid directory."
            return False

class Iso:

    def __init__(self, inputs, system):

        self.inputs = inputs
        self.system = system
        self.drive = self.system.wait_for_media()

    def build_list(self, files):

        disc_capacity = int(float(self.drive.capacity) * float(options.size_factor))
        print "Disc capacity reported by drive is: " + str(int(self.drive.capacity))
        if options.size_factor != 1:
            print " Adjusted disc capacity set to: " + str(disc_capacity)

        print "Building file lists..."
        file_count = 1
        # First 32768 bytes are unused by ISO 9660
        disc_size = 32768
        disc_list = []
        disc_files = []
        outfiles = []

        for file in files:

            if not os.path.islink(file):
            # Add ISO 9660 file system overhead in addition to file size
                file_size = 33 + len(os.path.split(file)[1]) + os.path.getsize(file)
                # Round up file sizes to 2KB sector size for ISO 9660.
                iso_size = math.ceil(file_size/(2.0**11))*(2.0**11)
                disc_size += iso_size
            else: pass

            if disc_size >= disc_capacity:
                disc_list.append(disc_files)
                print 'Disc %s will contain %s files using %dMB.' % (str(len(disc_list)), str(len(disc_files)),disc_size/(2**20))
                disc_size = 0
                disc_size += iso_size
                disc_files = []
                disc_files.append(file)
            else:
                disc_files.append(file)

        disc_list.append(disc_files)
        print 'Disc %s will contain %s files using %dMB.' % (str(len(disc_list)), str(len(disc_files)),disc_size/(2**20))
        return (disc_list)

    def calculate_discs(self):

        print "Calculating discs..."
        dir = self.inputs.backup_dir
        # First 32768 bytes are unused by ISO 9660
        total_size = 32768
        # Attempt to predict metadata size (file & directory descriptors)
        descriptor_size = 0
        disc_capacity = int(float(self.drive.capacity) * float(options.size_factor))

        if dir[len(dir)-1:] != "/":
            dir = dir + "/"

        file_list = []

        for path, dirs, files in os.walk(dir):
            for filename in files:
                file_list.append(os.path.join(path, filename))

        file_list.sort()
        for file in file_list:
            if not os.path.islink(file):
                size = os.path.getsize(file)
                if options.verbose: print file, size
                # Individual files must be smaller than 4GiB -1.
                if size >= 4*2**30 and not options.skip_big:
                    print "%s is to big. ISO 9660 requires individual files to be smaller than 4GiB." % file
                    sys.exit(1)
                elif size >= disc_capacity - 32768 and not options.skip_big:
                    print "%s will not fit on the current media. Try the --skip-big option to workaround." % file
                    sys.exit(1)
                elif size >= 4*2**30 and options.skip_big:
                    print "%s will be skipped because it is too big for ISO 9660." % file
                    file_list.remove(file)
                    continue
                elif size >= disc_capacity - 32768 and options.skip_big:
                    print "%s will be skipped because it is too big for the current media." % file
                    file_list.remove(file)
                    continue
                else:
                    total_size += size

        num_discs = int(math.ceil(total_size/disc_capacity))
        discs = self.build_list(file_list)
        print "\nNumber of %s's required to burn: %s" % ("dvd", len(discs))
        file_count = 0

        for disc in discs:
            file_count = file_count + len(disc)

        print "\nSanity Check:\n"
        print "Total files in directory: ", len(file_list)
        print "Total files in all discs: ", file_count

        print 'Total discs still: ' + str(len(discs))
        return discs


    def burn(self, disc, disc_num, total_disc, volume_name, test, iso_dir):


        msg = "\nReady to burn disc %s/%s."  % (disc_num, total_disc )
        print msg
        if int(disc_num) > 1:
            input=raw_input("Press enter to continue...")
        speed = self.inputs.speed
        drive = self.system.wait_for_media()
        dir = self.inputs.backup_dir

        list = ""
        for file in disc:

            file_on_disc = file.replace(dir, "")
            list = list + "%s=%s\n" % (file_on_disc, file)

        fd, temp_list = tempfile.mkstemp(suffix=".discspanlist")
        output = open(temp_list, 'w')
        output.write(list)
        output.close()

        if test :
            print 'Test write option enabled.'
            drive = '/dev/null'
        else:
            drive = self.drive.device

        if iso_dir :
            print 'ISO will be written to a file.'
            if iso_dir[len(iso_dir)-1:] != "/":
                iso_dir += "/"
            iso_file = iso_dir + volume_name + ".iso"
            burn_cmd = "mkisofs -o %s -V %s -A DiscSpan -p Unknown" \
                       " -iso-level 4 -l -r -hide-rr-moved -J -joliet-long" \
                       " -graft-points" % (iso_file, volume_name)
        else :
            burn_cmd = "growisofs -Z %s -speed=%s -use-the-force-luke=notray" \
                       " -use-the-force-luke=tty  " \
                       " -V %s -A DiscSpan -p Unknown -iso-level 4" \
                       " -l -r -hide-rr-moved -J -joliet-long" \
                       " -graft-points" % (drive, speed, volume_name)

        burn_cmd = burn_cmd + ' -path-list %s' % temp_list
        p = subprocess.Popen("%s" % (burn_cmd), shell=True)
        print burn_cmd
        sts = os.waitpid(p.pid, 0)
        os.unlink(temp_list)
        if not test and not iso_dir:
            p = subprocess.Popen("%s" % ("eject"), shell=True)
            sts = os.waitpid(p.pid, 0)
            sleep(5)

if __name__ == "__main__":

    parser = OptionParser(version=__version__)
    parser.add_option("--config",  dest="config_file", metavar ="CONFIG_FILE",
                      help="Location of config file.")
    parser.add_option("--start-disc", dest="start_disc",
                      default=1, help="Specify disc to start with (in case of failed previous burn)", metavar="START_DISC")
    parser.add_option("--skip-big", action="store_true", dest="skip_big",
                      default=False, help="Skip files that are too big.", metavar="SKIP_BIG")
    parser.add_option("--test", action="store_true", dest="test",
                      default=False, help="Performs a test run of the burn.", metavar="TEST")
    parser.add_option("--dir",  dest="backup_dir", metavar ="BACKUP_DIR",
                      help="Directory to backup.")
    parser.add_option("--volume-name",  dest="volume_name", metavar ="VOLUME_NAME",
                      default='DiscSpan', help="Name for the volume.")
    parser.add_option("--size-factor", dest="size_factor",
                      default=1, help="Specify size factor for disc capacity, i.e. 1 = 100%", metavar="SIZE_FACTOR")
    parser.add_option("--iso-dir",  dest="iso_dir", metavar ="ISO_DIR",
                      default=False, help="Redirect iso generation to a directory. Filename will be generated from volume name.")
    parser.add_option("-v", "--verbose", action="store_true", dest="verbose",
                      default=False, help="Extended verbosity", metavar="VERBOSE")

    (options, args) = parser.parse_args()

    n = 0

    config_file = options.config_file

    if config_file == None:

        for file in [ os.path.join(sys.path[0], 'discspan.ini'), '/usr/local/etc/discspan.ini', '/etc/discspan.ini'] :
            if os.path.isfile(file):
                config_file = file

    print "Using", config_file, "as config file."
    c = Config(config_file)
    s = System(c)
    interface = Interface(options.backup_dir,c.speed)
    iso = Iso(interface, s)
    discs = iso.calculate_discs()
    disc_num = int(options.start_disc)

    for disc in discs[(int(disc_num)-1):]:
            # Basic logic to prevent volume names that are too long.
        if len(options.volume_name) <= 30 - len(discs):
            volume_name = options.volume_name + '_' + str(disc_num)
        else: volume_name = options.volume_name

        iso.burn(disc, disc_num, len(discs), volume_name, options.test, options.iso_dir)
        disc_num = disc_num + 1
